[webkitdirs] Avoid call to `xcodebuild -showsdks` when computing Xcode options

emw-apple opened the pull request:
#### c6027549fd071239887c3b7fed40c73958f06119
<pre>
[webkitdirs] Avoid call to `xcodebuild -showsdks` when computing Xcode options
<a href="https://bugs.webkit.org/show_bug.cgi?id=243257">https://bugs.webkit.org/show_bug.cgi?id=243257</a>

Reviewed by Alexey Proskuryakov.

Since we only use the _name_ of the available SDKs, improve performance
by enumerating the SDK directories instead of calling xcodebuild.

This code is relatively hot compared to most build scripts, as it runs
at least once for every `make` invocation.

* Tools/Scripts/webkitdirs.pm:
(unversionedSDKNameFromSDK):
(availableXcodeSDKs):

Canonical link: <a href="https://commits.webkit.org/253224@main">https://commits.webkit.org/253224@main</a>
</pre>

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -645,8 +646,28 @@ sub parseAvailableXcodeSDKs($)
+
 sub availableXcodeSDKs
 {
+    # Looking for SDKs in known locations is much faster than calling through to xcodebuild.
```
Did you intend to commit this commented out?

emw-apple replied:
Oops, of course not 😛

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -38,6 +38,7 @@ use Digest::MD5 qw(md5_hex);
 use FindBin;
 use File::Basename;
 use File::Find;
+use File::Glob qw(bsd_glob);
```
Hoping that this won't break any platforms by itself.

emw-apple replied:
`bsd_glob` isn't OS-specific, it's just behaves more like modern POSIX glob than perl's `glob` function does. The main reason I'm using it is to support SDK paths containing whitespace (though I suspect other parts of our build process do not). From perl's docs, emphasis mine:

> File::Glob::bsd_glob() implements the FreeBSD glob(3) routine, which is a superset of the POSIX glob() (described in IEEE Std 1003.2 "POSIX.2"). bsd_glob() takes a mandatory pattern argument, and an optional flags argument, and returns a list of filenames matching the pattern, with interpretation of the pattern modified by the flags variable.
>
> Since v5.6.0, Perl's CORE::glob() is implemented in terms of bsd_glob(). Note that they don't share the same prototype--CORE::glob() only accepts a single argument. **Due to historical reasons, CORE::glob() will also split its argument on whitespace, treating it as multiple patterns, whereas bsd_glob() considers them as one pattern**. But see :bsd_glob under ["EXPORTS"](https://perldoc.perl.org/File::Glob#EXPORTS), below.

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -645,8 +646,28 @@ sub parseAvailableXcodeSDKs($)
+
 sub availableXcodeSDKs
 {
+    # Looking for SDKs in known locations is much faster than calling through to xcodebuild.
```
Is there a radar for making it faster? Maybe it would also explain why it has to be slow, and what's wrong with this optimization :)

emw-apple replied:
Filed rdar://98220832.

aproskuryakov replied:
Thank you. It would be optimal to have a link in the comment, so that anyone (an Apple employee) could immediately see if the workaround was still necessary in a few years from now.

emw-apple committed:
[webkitdirs] Avoid call to `xcodebuild -showsdks` when computing Xcode options
https://bugs.webkit.org/show_bug.cgi?id=243257

Reviewed by Alexey Proskuryakov.

Since we only use the _name_ of the available SDKs, improve performance
by enumerating the SDK directories instead of calling xcodebuild.

This code is relatively hot compared to most build scripts, as it runs
at least once for every `make` invocation.

* Tools/Scripts/webkitdirs.pm:
(unversionedSDKNameFromSDK):
(availableXcodeSDKs):

Canonical link: https://commits.webkit.org/253224@main
